File: AstAnalyzer.js
// New file: c:\Users\Tim\Desktop\LMS\astAnalyzer.js
const parser = require('@babel/parser');
const traverse = require('@babel/traverse').default;

class AstAnalyzer {
  constructor() {
    this.parseOptions = {
      sourceType: 'module',
      plugins: ['jsx', 'typescript', 'classProperties', 'decorators-legacy']
    };
  }
  
  parseToAst(content, filePath = 'unknown') {
    try {
      return parser.parse(content, this.parseOptions);
    } catch (error) {
      return null; // Return null on parsing error
    }
  }
  
  calculateComplexity(ast) {
    if (!ast) return 1;
    
    try {
      let complexity = 1;
      
      // Rest of method...
      
      console.log(`Debug: Calculated complexity: ${complexity} for AST`);
      return complexity;
    } catch (error) {
      console.warn(`Error calculating complexity: ${error.message}`);
      return 1;
    }
  }
  
  analyzeComponentAst(content, filePath) {
    // Move existing analyzeComponentAst method here
  }

  /**
   * Estimate complexity for Rust code (basic version)
   */
  estimateRustComplexity(content) {
    let complexity = 1;
    // Count decision points: if, else if, for, while, loop, match arms, ? operator
    complexity += (content.match(/if\s|else\s+if|for\s|while\s|loop\s*\{|match\s|=>|Ok\s*\(/g) || []).length;
    // Add complexity for closures
    complexity += (content.match(/\|[^|]*\|/g) || []).length;
    // Add complexity for nested functions (basic check)
    complexity += (content.match(/fn\s+\w+\s*\(/g) || []).length - 1; // Subtract 1 for the main function if present

    return Math.max(1, complexity); // Ensure complexity is at least 1
  }
}

module.exports = AstAnalyzer;

File: src/analyzers/modules/ast_analyzer.rs
use swc_ecma_ast::*;
use swc_ecma_parser::{lexer::Lexer, Parser, StringInput, Syntax};
use swc_common::{
    sync::Lrc,
    SourceMap,
    errors::{ColorConfig, Handler},
};
use std::path::Path;

/// AST Analyzer for JavaScript/TypeScript code analysis
pub struct AstAnalyzer {
    source_map: Lrc<SourceMap>,
    handler: Handler,
}

impl AstAnalyzer {
    /// Create a new AST analyzer instance
    pub fn new() -> Self {
        let source_map = Lrc::new(SourceMap::default());
        let handler = Handler::with_tty_emitter(ColorConfig::Auto, true, false, Some(source_map.clone()));
        
        Self {
            source_map,
            handler,
        }
    }
    
    /// Parse JavaScript/TypeScript content to AST
    pub fn parse_to_ast(&self, content: &str, file_path: &str) -> Option<Module> {
        let fm = self.source_map.new_source_file(
            swc_common::FileName::Custom(file_path.into()),
            content.into(),
        );
        
        let lexer = Lexer::new(
            Syntax::Es(Default::default()),
            Default::default(),
            StringInput::from(&*fm),
            None,
        );
        
        let mut parser = Parser::new_from(lexer);
        
        match parser.parse_module() {
            Ok(module) => Some(module),
            Err(_) => {
                // Return None on parsing error
                None
            }
        }
    }
    
    /// Calculate code complexity for an AST
    pub fn calculate_complexity(&self, ast: &Option<Module>) -> usize {
        match ast {
            None => 1,
            Some(module) => {
                let mut complexity = 1;
                
                // Count items in the module
                complexity += module.body.len();
                
                // Additional complexity analysis would go here
                // This would include traversing the AST and counting:
                // - if statements and branches
                // - loops
                // - function definitions
                // - etc.
                
                complexity
            }
        }
    }
    
    /// Analyze component AST (placeholder for future implementation)
    pub fn analyze_component_ast(&self, content: &str, file_path: &str) -> Option<ComponentAnalysis> {
        // Placeholder implementation
        Some(ComponentAnalysis {
            component_name: Path::new(file_path).file_stem()?.to_string_lossy().to_string(),
            complexity: self.estimate_complexity(content),
            props: vec![],
            dependencies: vec![],
        })
    }
    
    /// Estimate complexity for JavaScript code
    pub fn estimate_complexity(&self, content: &str) -> usize {
        let mut complexity = 1;
        
        // Count decision points: if, else if, for, while, switch cases
        let if_count = content.matches("if ").count();
        let else_if_count = content.matches("else if").count();
        let for_count = content.matches("for ").count();
        let while_count = content.matches("while ").count();
        let switch_case_count = content.matches("case ").count();
        
        complexity += if_count + else_if_count + for_count + while_count + switch_case_count;
        
        // Add complexity for function definitions
        complexity += content.matches("function ").count();
        
        complexity
    }
    
    /// Estimate complexity for Rust code
    pub fn estimate_rust_complexity(&self, content: &str) -> usize {
        let mut complexity = 1;
        
        // Count decision points: if, else if, for, while, loop, match arms, ? operator
        let if_count = content.matches("if ").count();
        let else_if_count = content.matches("else if").count();
        let for_count = content.matches("for ").count();
        let while_count = content.matches("while ").count();
        let loop_count = content.matches("loop {").count();
        let match_count = content.matches("match ").count();
        let question_mark_count = content.matches("?").count();
        
        complexity += if_count + else_if_count + for_count + while_count + loop_count + match_count + question_mark_count;
        
        // Add complexity for closures
        complexity += content.matches("|").count() / 2; // Rough approximation
        
        // Add complexity for nested functions (basic check)
        let fn_count = content.matches("fn ").count();
        if fn_count > 0 {
            complexity += fn_count - 1; // Subtract 1 for the main function if present
        }
        
        complexity
    }
}

/// Represents analysis of a UI component
pub struct ComponentAnalysis {
    pub component_name: String,
    pub complexity: usize,
    pub props: Vec<PropInfo>,
    pub dependencies: Vec<String>,
}

/// Information about a component prop
pub struct PropInfo {
    pub name: String,
    pub prop_type: String,
    pub required: bool,
}
